core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2124,12 +2132,14 @@ class Log(@volatile private var _dir: File,
           time = time,
           initFileSize = initFileSize,
           preallocate = config.preallocate))
-        updateLogEndOffset(newOffset)
         leaderEpochCache.foreach(_.clearAndFlush())
 
-        producerStateManager.truncate()
-        producerStateManager.updateMapEndOffset(newOffset)
+        // Clear producer state prior to adjusting log start and end offsets
+        // in case the first unstable offset is invalidated by the truncation.
+        producerStateManager.truncateFullyAndStartAt(newOffset)
         maybeIncrementFirstUnstableOffset()
+
+        updateLogEndOffset(newOffset)
         updateLogStartOffset(newOffset)
```

Comment:
One note here: in the initial patch, I reordered `updateLogStartOffset` before `updateLogEndOffset`. This caused the  test case `ReassignPartitionsIntegrationTest.testHighWaterMarkAfterPartitionReassignment` to fail. The problem is the logic in `updateHighWatermark`, which is called in `updateLogStartOffset`, assumes that the log end offset has already been set correctly. The result of reversing the order was that the high watermark was left at a value lower than the log start offset. 

Frankly the test case is a bit contrived. It calls `truncateFullyAndStartAt` on the leader before starting a reassignment. I don't think there are any valid operations that could have the same effect. At the same time, it feels brittle to rely on tricky ordering assumptions like this, so I'm wondering whether we can improve this. One idea for the case of `truncateFullyAndStartAt` is to bypass `updateLogEndOffset` and `updateLogStartOffset` and update the values directly. Any thoughts?

Reply:
Maybe we can have a new method that updates the producerState, logEndOffset, high watermark, log start offset and last stable offset together? Sth like the following.

```
    rebuildProducerState()
    logStartOffset = offset
    nextOffsetMetadata = LogOffsetMetadata(offset, activeSegment.baseOffset, activeSegment.size)

    // Update the high watermark metadata if the offset changes
    if (highWatermark != offset) {
      updateHighWatermarkMetadata(nextOffsetMetadata)
    }

    if (this.recoveryPoint != offset) {
      this.recoveryPoint = offset
    }

```

This avoids dependencies and also avoids some the the duplication of the logic that Chia-Ping mentioned.

Reply:
Thanks. I tried something like this in the latest commit. Let me know what you think.